include/nsysnet/nssl.h

```diff
@@ -4,46 +4,318 @@
+    NSSL_SERVER_CERT_COMODO_RSA_CA = 1900,
+    //! USERTrust RSA Certification Authority
+    //! (SHA1: 2B 8F 1B 57 33 0D BB A2 D0 7A 6C 51 F7 0E E9 0D DA B9 AD 8E)
+    NSSL_SERVER_CERT_USERTRUST_RSA_CA = 1901,
+    //! Represents the last certificate in the 4096-bit commercial CA group.
+    NSSL_SERVER_CERT_GROUP_2_LAST = 1901,
 } NSSLServerCertId;
 
+/**
+ * Initialises the NSSL library for use. This must be called before any other
+ * NSSL functions.
+ *
+ * \returns
+ * 0 on success, or a negative error code on failure.
+ *
+ * \sa
+ * - NSSLFinish()
+ */
 int32_t
 NSSLInit();
 
+/**
+ * Cleans up and finishes use of the NSSL library.
+ *
+ * \returns
+ * 0 on success, or -1 on error.
+ */
 int32_t
 NSSLFinish();
 
+/**
+ * Creates a context to use with the NSSL library. A context contains root CAs
+ * and connections.
+ *
+ * \param unk
+ * Unknown. A value of 0 appears to work.
+ *
+ * \returns
+ * A #NSSLContextHandle representing the newly created context, or a negative
+ * value on error.
+ *
+ * \sa
+ * - NSSLDestroyContext()
+ */
 NSSLContextHandle
 NSSLCreateContext(int32_t unk);
 
+/**
+ * Destroys a NSSL context.
+ *
+ * \param context
+ * The NSSL context to destroy.
+ *
+ * \returns
+ * 0 on success, or a negative value if an error occured.
+ */
 int32_t
 NSSLDestroyContext(NSSLContextHandle context);
 
+/**
+ * Add a custom root CA to the given NSSL context.
+ *
+ * \param context
+ * The NSSL context to add a root CA to.
+ *
+ * \param cert
+ * A pointer to a buffer containing the certificate data.
+ *
+ * \param length
+ * The length of the certificate buffer.
+ *
+ * \param unk
+ * Unknown.
+ *
+ * \returns
+ * 0 on success, or a negative value on error.
+ *
+ * \sa
+ * - NSSLAddServerPKI()
+ *
+ * \if false
+ * meta: what format does the certificate need to be? The internal ones are
+ * x509 in binary DER. What's a working value for unk? 0?
+ * \endif
+ */
 int32_t
 NSSLAddServerPKIExternal(NSSLContextHandle context,
                          const void *cert,
                          int32_t length,
                          int32_t unk);
 
+/**
+ * Add a system root CA (one of #NSSLServerCertId) to the given NSSL context.
+ *
+ * \param context
+ * The NSSL context to add a root CA to.
+ *
+ * \param pki
+ * The #NSSLServerCertId representing the system CA to add.
+ *
+ * \returns
+ * 0 on success, or a negative value on error.
+ *
+ * \sa
+ * - NSSLAddServerPKIExternal()
+ */
 int32_t
 NSSLAddServerPKI(NSSLContextHandle context,
                  NSSLServerCertId pki);
 
+/**
+ * Open an SSL connection.
+ *
+ * \param context
+ * The context to open the connection under.
+ *
+ * \param host
+ * The domain or IP of the host to connect to.
+ *
+ * \param hostLength
+ * The length of the "host" buffer.
+ *
+ * \param options
+ * Options for the underlying socket.
+ *
+ * \param socket
+ * An existing socket to use for the connection, or NULL.
+ *
+ * \param block
+ * Whether to block until the connection is open.
```

Comment:
This probably means whether the socket is in blocking mode, so will block on all operations (read, write, etc) not just open.